src/axom/primal/geometry/Quadrilateral.hpp

```diff
@@ -0,0 +1,264 @@
+   *
+   * \note It is the responsiblity of the caller to pass
+   *       an array with at least 4 Points
+   */
+  AXOM_HOST_DEVICE
+  explicit Quadrilateral(const PointType* pts)
+  {
+    for(int i = 0; i < NUM_QUAD_VERTS; i++)
+    {
+      m_points[i] = pts[i];
+    }
+  }
+
+  /*!
+   * \brief Quadrilateral constructor from an Array of Points.
+   *
+   * \param [in] pts An ArrayView containing 4 Points.
+   */
+  AXOM_HOST_DEVICE
+  explicit Quadrilateral(const axom::ArrayView<PointType> pts)
+  {
+    SLIC_ASSERT(pts.size() == NUM_QUAD_VERTS);
+
+    for(int i = 0; i < NUM_QUAD_VERTS; i++)
+    {
+      m_points[i] = pts[i];
+    }
+  }
+
+  /*!
+   * \brief Quadrilateral constructor from an initializer list of Points
+   *
+   * \param [in] pts an initializer list containing 4 Points
+   */
+  AXOM_HOST_DEVICE
+  explicit Quadrilateral(std::initializer_list<PointType> pts)
+  {
+    SLIC_ASSERT(pts.size() == NUM_QUAD_VERTS);
+
+    int i = 0;
+    for(const auto& pt : pts)
+    {
+      m_points[i] = pt;
+      i++;
+    }
+  }
+
+  /*!
+   * \brief Index operator to get the i^th vertex
+   * \param idx The index of the desired vertex
+   * \pre idx is 0, 1, 2, or 3
+   */
+  AXOM_HOST_DEVICE
+  PointType& operator[](int idx)
+  {
+    SLIC_ASSERT(idx >= 0 && idx < NUM_QUAD_VERTS);
+    return m_points[idx];
+  }
+
+  /*!
+   * \brief Index operator to get the i^th vertex
+   * \param idx The index of the desired vertex
+   * \pre idx is 0, 1, 2, or 3
+   */
+  AXOM_HOST_DEVICE
+  const PointType& operator[](int idx) const
+  {
+    SLIC_ASSERT(idx >= 0 && idx < NUM_QUAD_VERTS);
+    return m_points[idx];
+  }
+
+  /// \brief Returns the area of the quadrilateral (2D specialization)
+  template <int TDIM = NDIMS>
+  AXOM_HOST_DEVICE typename std::enable_if<TDIM == 2, double>::type area() const
+  {
+    return axom::utilities::abs(signedArea());
+  }
+
+  /**
+   * \brief Returns the signed area of a 2D quadrilateral
+   *
+   * The area is positive when the vertices are oriented counter-clockwise.
+   * \note This function is only available for quadrilaterals in 2D since
+   * signed areas don't make sense for 3D quadrilaterals
+   *
+   * Compute the signed area by dividing the quadrilateral into two triangles
+   * as shown below and summing their signed area
+   *
+   * \verbatim
+   *
+   * D +----+ C       +y
+   *   |   /|         
+   *   |  / |          ^
+   *   | /  |          |
+   *   |/   |          |
+   * A +----+ B        -----> +x
+   *
+   * \endverbatim
+   */
+  template <int TDIM = NDIMS>
+  AXOM_HOST_DEVICE typename std::enable_if<TDIM == 2, double>::type signedArea() const
+  {
+    const PointType& A = m_points[0];
+    const PointType& B = m_points[1];
+    const PointType& C = m_points[2];
+    const PointType& D = m_points[3];
+
+    // clang-format off
+    return TriangleType{A, B, C}.signedArea() +
+           TriangleType{C, D, A}.signedArea();
+    // clang-format on
+  }
```

Comment:
Can we add an `area` function in 3D too?
It's area would be
```cpp
    return TriangleType{A, B, C}.area() +
           TriangleType{C, D, A}.area();
```

Reply:
Possible optimization for the future: We might be able to find/devise a way to directly compute the area without breaking the quad into two triangles, similarly to how we compute the volume of a hexahedron.

Reply:
> Can we add an `area` function in 3D too? It's area would be

Thinking about it a bit more, 3D area is more complicated if the quad is non-planar.
For now, I'll "unask" the question and wait for an actual need.